### ISSUE TITLE
Mergeable funnels

### ISSUE THREAD

**Issue by carlomanf** [carlomanf/wp-funnel-manager#18]
The current implementation of funnel steps requires each step to belong to one funnel only, but allowing steps to be shared between funnels could prove useful.

Consider the following example, where A to G are funnel steps.

```
A->B->F->G
C->D->E->F->G
```

In the example, the two funnels headed by A and C merge into one, because both B and E flow into F, with F then flowing into G.

One idea to implement this is to internally step up three funnel structures, with each one indicating a tail via its head:

```
(A tail F)->B
(C tail F)->D->E
(F no tail)->G
```

Especially after #15 is implemented, a post status or similar would also be necessary to prevent funnel F from being exposed in its own right. (also see #10)

This new feature represents an opportunity to address a number of long-standing issues with the plugin, such as #9, #10, #4 and #3. Specifically:

- #3 may not be as much of an issue when it is possible to share steps between funnels.
- An internal implementation that makes use of "tails" could potentially aid with a more controlled and logical re-ordering experience for #4, rather than the ability to re-order without restriction, which is arguably not as helpful as it may first seem.
- #9 and #10 are caused by the funnel head being implemented as a separate post type to the interior steps. This partially helps with the aforementioned re-ordering experience, by preventing the head from changing, but may no longer be needed.

**Comment by carlomanf** [comment]
One thing that needs to be determined is how compatible this feature will be with the existing ways that funnels are built. While it will definitely not need a whole new funnel type like #11 and #15, it is yet to be determined whether it will be feasible to merge funnels of the older legacy types or whether mergeability must be reserved for the modern funnel type.

**Comment by carlomanf** [comment]
The example over at #25 raises an important consideration for mergeable funnels.

> The most useful purpose of mergeable funnels is that of conversion testing. Suppose that for every visitor that reaches step F is worth $5. If 100 visitors start at A and 5 of them reach F, then A is worth $0.25 by extrapolation. If a different 100 visitors start at C and 3 of them reach F, then C is worth $0.15 by extrapolation.

The example is fairly straightforward assuming that the 100 starting at A and the 100 starting at C are disjoint. If, however, there are visitors who visited both funnels A and C, the conversion data can get muddled.

Particularly, if a visitor reaches G via C and F, access to steps F and G become both unlocked. If the same visitor were then to visit A and be allowed to skip ahead to G within that funnel without first re-visiting F, it would not necessarily be correct to report that step F is converting at 80% like in the example in #25. In an extreme case, without the appropriate safeguards, step F could be incorrectly reported to convert above 100%.

The access control of mergeable funnels should be carefully implemented so as to avoid spurious conversion reports when that feature is later implemented.

**Comment by carlomanf** [comment]
Ideally the above issue would be solved without requiring visitors to re-visit steps they have already visited. I believe it can be, provided the following:

1. Visitors can skip steps within the merged section but not steps preceding the merged section. For example, assuming a visitor has already visited `A->B->F->G`, then `C->D->E->G` would be allowed but not `C->G` or `C->D->G` or `C->F` or `C->D->F`.
2. When calculating the conversion report, as soon as a visitor reaches any step within a merged section, they are added to the traffic total of all preceding steps within the merged section, even if they only ever visited those steps through a different funnel.

Implementing number 2 is relevant to #25 and not to this issue, but number 1 will also be required to ensure that the step immediately preceding a merged section (e.g. E) does not display an incorrectly high conversion rate.

**Comment by carlomanf** [comment]
Something not yet discussed is the user experience for creating a merged funnel through the admin.

For the initial implementation in v1.5.0, I would like to keep it simple and provide the following creation gateways:

1. A "Merge" bulk action for funnels that requires multiple funnels to be selected, creates a tail for them to merge into and redirects to the step listing for the tail.

2. A "Merge New Funnel" action that appears per step, creates a new funnel that merges into that step, creates a tail to contain the merged steps and redirects to the funnel editor.

**Comment by carlomanf** [comment]
The basic structure of this is now introduced as of b33ba03. Further enhancements and bug fixes can be tracked in new issues.
